### PR TITLE
TST DOC Window docs, tests, etc.

### LINKED ISSUE #4234: scipy.signal.flattop documentation doesn't list a source for the function
There are a whole bunch of flattop window functions for use in signal analysis, a "flattop" isn't descriptive enough, and another source isn't listed.  Does anyone know where this function is from?

### PR DESCRIPTION
**Pull request by endolith**
This should fix https://github.com/scipy/scipy/issues/4234 and improve coverage of exceptions, etc. shown by codecov.

Combined several window functions into a generic sum-of-cosines function, since that's how most references describe the different functions anyway:

> This kind of optimization leads to the definition of the so-called flat-top windows. If a
> third order window is considered, this optimization yields the following coefficients: A<sub>0</sub> = 0.2811, A<sub>1</sub> = -0.5209, A<sub>2</sub> = 0.1980.

So now you can reproduce this window with  `_cos_win(M, [0.2811, 0.5209, 0.1980])`.

### CODE REVIEW

File: scipy/signal/windows.py [hunk 1]
```diff
@@ -693,8 +745,8 @@ def hann(M, sym=True):
         return np.ones(M)
     M, needs_trunc = _extend(M, sym)
 
-    n = np.arange(0, M)
-    w = 0.5 - 0.5 * np.cos(2.0 * np.pi * n / (M - 1))
+    a = [0.5, 0.5]
+    w = _cos_win(M, a)
```
Comment (Eric89GXL) [inline scipy/signal/windows.py]: With all of these it's probably worth saving a few lines with the equally readable `w = _cos_win(M, a=[...])`

File: scipy/signal/tests/test_windows.py [hunk 2]
```diff
@@ -80,7 +80,7 @@ def test_cheb_odd_high_attenuation(self):
     def test_cheb_even_high_attenuation(self):
         with warnings.catch_warnings():
             warnings.simplefilter("ignore", UserWarning)
-            cheb_even = signal.chebwin(54, at=-40)
+            cheb_even = signal.chebwin(54, at=40)
```
Comment (Eric89GXL) [inline scipy/signal/tests/test_windows.py]: Were these changes necessary to get the tests to pass, or just for aesthetics?

**Comment by Eric89GXL** [comment]
Otherwise LGTM

File: scipy/signal/tests/test_windows.py (hunk 2 above)
Comment (endolith) [inline scipy/signal/tests/test_windows.py]: For codecov, the line calls `abs(at)` and positive values of `at` weren't being tested.

File: scipy/signal/windows.py (hunk 1 above)
Comment (endolith) [inline scipy/signal/windows.py]: ok

File: scipy/signal/tests/test_windows.py (hunk 2 above)
Comment (Eric89GXL) [inline scipy/signal/tests/test_windows.py]: Sounds good

**Comment by Eric89GXL** [comment]
Ping me when the last comment is addressed and the CIs are happy and I'll go ahead and merge.

**Comment by endolith** [comment]
@Eric89GXL Do you still have Matlab access?  Can you get the output of all of its window functions for odd- and even-length, both symmetric and periodic, so we can check consistency with it?  I think length 6 and 7 are enough to distinguish them without making the test file terribly long?  (5 is not: `bartlett(5) == hann(5) == barthann(5)`)

I think something like this would work:

```
clear all; close all; clc
format long e
format compact

barthannwin(6)
barthannwin(7)
bartlett(6)
bartlett(7)
blackman(6, 'periodic')
blackman(7, 'periodic')
blackman(6, 'symmetric')
blackman(7, 'symmetric')
blackmanharris(6, 'periodic')
blackmanharris(7, 'periodic')
blackmanharris(6, 'symmetric')
blackmanharris(7, 'symmetric')
bohmanwin(6)
bohmanwin(7)
chebwin(6, 100)
chebwin(7, 100)
chebwin(6, 10)
chebwin(7, 10)
flattopwin(6, 'periodic')
flattopwin(7, 'periodic')
flattopwin(6, 'symmetric')
flattopwin(7, 'symmetric')
gausswin(6)
gausswin(7)
gausswin(6, 1)
gausswin(7, 1)
hamming(6, 'periodic')
hamming(7, 'periodic')
hamming(6, 'symmetric')
hamming(7, 'symmetric')
hann(6, 'periodic')
hann(7, 'periodic')
hann(6, 'symmetric')
hann(7, 'symmetric')
kaiser(6)
kaiser(7)
kaiser(6, 2.7)
kaiser(7, 2.7)
nuttallwin(6, 'periodic')
nuttallwin(7, 'periodic')
nuttallwin(6, 'symmetric')
nuttallwin(7, 'symmetric')
parzenwin(6)
parzenwin(7)
rectwin(6)
rectwin(7)
taylorwin(6)
taylorwin(7)
taylorwin(6, 5, -25)
taylorwin(7, 5, -25)
triang(6)
triang(7)
tukeywin(6, 0)
tukeywin(7, 0)
tukeywin(6, .25)
tukeywin(7, .25)
tukeywin(6)
tukeywin(7)
tukeywin(6, .75)
tukeywin(7, .75)
tukeywin(6, 1)
tukeywin(7, 1)
```

File: scipy/signal/windows.py (hunk 1 above)
Comment (endolith) [inline scipy/signal/windows.py]: ok I shortened most of them except one that was too long for a single line that way.

could also put the guards and stuff into the _cos_win function so that blackman becomes just 

    return _cos_win(M, a=[0.42, 0.50, 0.08], sym)

for instance.

**Comment by e-q** [comment]
Here is the output of what you wrote: 
https://gist.github.com/e-q/6cdaf9b098b76a1c2e428e6b583ec999

This looks like a great PR!

**Comment by endolith** [comment]
@e-q 

Thanks!  What's different about https://gist.github.com/Eric89GXL/8239281 that it has so many more digits though? I thought if I copied the beginning lines it would do the same thing.

**Comment by e-q** [comment]
Hm, maybe my local MATLAB config messed with it. I tried again, and it looks like it came out right.

**Comment by endolith** [comment]
@e-q Matlab's `triang(6 or 7)` really goes above 1?  o_O

**Comment by Eric89GXL** [comment]
Not on my system, 2013a:
```
>> triang(6)
ans =
    0.1667
    0.5000
    0.8333
    0.8333
    0.5000
    0.1667
>> triang(7)
ans =
    0.2500
    0.5000
    0.7500
    1.0000
    0.7500
    0.5000
    0.2500
```

**Comment by e-q** [comment]
I don't think so. Maybe you miscounted the outputs and are looking at the `taylorwin(6, 5, -25)` and `taylorwin(7, 5, -25)` calls?

**Comment by endolith** [comment]
Oops, yes. 😬  Still weird that `taylorwin` isn't normalized to 1, but it's a frequency domain window and we don't have it yet, so I won't worry about it.

**Comment by endolith** [comment]
I think I will leave fixing https://github.com/scipy/scipy/issues/2190 for another PR in case it gets delayed, etc.

**Comment by Eric89GXL** [comment]
Great. LGTM +1 for merge. @e-q do you have time to have a look and merge if you're happy?

**Comment by endolith** [comment]
Added one more ref, squashed some stuff and rebased

**Comment by e-q** [comment]
Very nice @endolith, thanks!
